atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java

```diff
@@ -504,11 +503,23 @@ public void notifySnapshotReplicationCompleted() {
   /**
    * Compacts the server logs.
    *
+   * <p>TODO: the resulting future is completed on the Raft thread, so make sure callbacks for it
```

Comment:
:question: Why is this a todo comment?

:wrench: I think the comment can be slightly misleading, see https://stackoverflow.com/questions/46060438/in-which-thread-do-completablefutures-completion-handlers-execute for context. Registered callbacks (e.g. just a `whenComplete`/`thenApply`) are unfortunately not guaranteed to run on the raft thread. I think that's important to mention so that we don't accidentally start using it under the assumption that a callback is safe to access the raft context.

Reply:
Is it not? Since the future is completed in the Raft thread, won't it trigger the callback in a blocking fashion, unless you called them with their `*Async` variants?

> Actions supplied for dependent completions of non-async methods may be performed by the thread that completes the current CompletableFuture, or by any other caller of a completion method.

In this case, the future is completed in the Raft thread. In which cases would the callback enqueued through a non-async method be called in a different thread?

Reply:
Nevermind, I get it. What a weird way of doing things :smile: I would not have expected a separate callback to execute the complete pipeline, but only itself and descendants.